src/qmapshack/poi/CPoiPOI.cpp

```diff
@@ -57,9 +66,128 @@ void CPoiPOI::draw(IDrawContext::buffer_t& buf)
+
+    for(QString category : tagMap.keys())
+    {
+        QStringList parentChild = category.split("_");
+        QString parent = parentChild[0];
+        QString child = parentChild[1];
+
+        auto parentItem = parentMap.value(parent, new QTreeWidgetItem());
+        parentItem->setData(CPoiPropSetup::eTreeColumnCategory, Qt::DisplayRole, parent);
+        parentItem->setCheckState(CPoiPropSetup::eTreeColumnCheckbox, Qt::Unchecked);
+        parentMap[parent] = parentItem;
+        QTreeWidgetItem* item = new QTreeWidgetItem(parentItem);
+        item->setData(CPoiPropSetup::eTreeColumnParentCategory, Qt::DisplayRole, parent);
+        item->setData(CPoiPropSetup::eTreeColumnCategory, Qt::DisplayRole, child);
+        item->setCheckState(CPoiPropSetup::eTreeColumnCheckbox, categoryActivated[category] ? Qt::Checked : Qt::Unchecked);
+    }
+    topLevelItems = parentMap.values();
+}
+
+void CPoiPOI::checkedStateChanged(QTreeWidgetItem * item, int columnChanged)
+{
+    if(columnChanged == CPoiPropSetup::eTreeColumnCheckbox)  //shouldn't happen, but still
+    {
+        QString parent = item->data(CPoiPropSetup::eTreeColumnParentCategory, Qt::DisplayRole).toString();
+        if(parent == "")
+        {
+            //This is a parent category
+            return;
+        }
+        QString child = item->data(CPoiPropSetup::eTreeColumnCategory, Qt::DisplayRole).toString();
+        categoryActivated[parent + "_" + child] = item->checkState(CPoiPropSetup::eTreeColumnCheckbox) == Qt::Checked;
+
+        loadTimer.start();
+    }
+}
+
+void CPoiPOI::loadPOIsFromFile(QString category, int minLonM10, int minLatM10)
+{
+    //Open Database here so it is owned by the right thread
+    if(!QSqlDatabase::contains(filename))
+    {
+        QSqlDatabase db = QSqlDatabase::addDatabase("QSQLITE", filename);
+        db.setDatabaseName(filename);
+        if(!db.open())
+        {
+            qDebug() << "failed to open database" << db.lastError();
+            isActivated = false;
+            return;
+        }
+    }
+
+    const QString queryString = "SELECT main.poi_index.maxLat, main.poi_index.maxLon, main.poi_index.minLat, main.poi_index.minLon, main.poi_data.data"
+                                " FROM main.poi_index"
+                                " INNER JOIN main.poi_data ON main.poi_index.id=main.poi_data.id"
+                                " WHERE main.poi_index.maxLat<" + QString::number((minLatM10 + 1) / 10., 'f') +
+                                " AND main.poi_index.minLat>" + QString::number(minLatM10 / 10., 'f') +
+                                " AND main.poi_index.maxLon<" + QString::number((minLonM10 + 1) / 10., 'f') +
+                                " AND main.poi_index.minLon>" + QString::number(minLonM10 / 10., 'f') +
+                                " AND main.poi_data.data LIKE '%" + tagMap[category] + "%'";
+
+    QSqlQuery query(queryString, QSqlDatabase::database(filename));
```

Comment:
Have a look into `CDBItem.cpp`. Using `bind()` is much safer to get a well formatted SQL query.

```
    QSqlQuery query(db);
    query.prepare("SELECT trash FROM items WHERE id=:id");
    query.bindValue(":id", id);
    QUERY_EXEC(return );
```

Reply:
Do you have an idea on how to reformat that also for the string that comes from tagMap? The problem is, that the query has a different format depending on the category

Reply:
hm, ok I see your point.  Don't have a good solution for that right at hand.